File: claude-craft/app/access.py
"""模块权限划分：自研访问控制层。

以「角色(module role)」为最小授权单元，通过策略表决定某角色能否调用某能力。
网关(gateway)执行前一律过 guard()：无权限即抛 AccessDenied，阻断越域调用。
本层独立于引擎与 UI，保证任何入口(HTTP/测试)都不绕过授权检查。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable

from app.config import Role


class AccessDenied(PermissionError):
    """越权访问被拒绝。"""

    def __init__(self, capability: str, role: str) -> None:
        self.capability = capability
        self.role = role
        super().__init__(f"角色 `{role}` 无权调用能力 `{capability}`")


@dataclass
class PermissionPolicy:
    """策略表:能力 -> 允许的角色集合。空集合=任何角色都不可用。"""

    grants: dict[str, set[str]] = field(default_factory=dict)

    def allow(self, capability: str, *roles: str) -> None:
        self.grants.setdefault(capability, set()).update(roles)

    def bulk_grants(self, table: dict[str, list[str]]) -> None:
        for cap, roles in table.items():
            self.allow(cap, *roles)


class AccessControl:
    """授权守卫。每个 module 持有其身份 role，调用能力前传 identity 校验。"""

    def __init__(self, policy: PermissionPolicy) -> None:
        self._policy = policy

    def grant(self, capability: str, *roles: str) -> None:
        self._policy.allow(capability, *roles)

    def bulk_grants(self, table: dict[str, list[str]]) -> None:
        self._policy.bulk_grants(table)

    def can(self, role: str, capability: str) -> bool:
        allowed = self._policy.grants.get(capability, set())
        return role in allowed

    def guard(self, role: str, capability: str) -> None:
        if not self.can(role, capability):
            raise AccessDenied(capability, role)


def grant_guard(role: str, ac: AccessControl):
    """装饰器:包一层带固定身份的打点函数，供注入使用。"""

    def decorate(capability: str):
        return lambda: ac.guard(role, capability)

    return decorate


def default_policy() -> PermissionPolicy:
    """默认授权:工具执行仅 TOOL;引擎仅 CORE;网关读写仅 GATEWAY;记忆仅 MEMORY。"""
    p = PermissionPolicy()
    table = {
        # 能力名                    -> 可调用角色
        "engine.run":               [Role.CORE],
        "engine.plan":              [Role.CORE],
        "tool.fs_read":             [Role.TOOL, Role.CORE],
        "tool.fs_write":            [Role.TOOL],
        "tool.fs_list":             [Role.TOOL, Role.CORE],
        "tool.sh_run":              [Role.TOOL],
        "memory.latch":             [Role.MEMORY, Role.CORE],
        "memory.invalidate":        [Role.MEMORY],
        "gateway.threads.read":     [Role.GATEWAY],
        "gateway.threads.write":    [Role.GATEWAY],
        "gateway.stream":           [Role.GATEWAY],
        "ui.emit":                  [Role.CORE, Role.UI],
    }
    p.bulk_grants(table)
    return p


__all__ = ["AccessControl", "AccessDenied", "PermissionPolicy", "default_policy", "grant_guard"]

File: server/src/engine/access.ts
/**
 * access.ts —— 自研角色/能力授权守卫。
 *
 * 以「角色」为授权最小单元：任意入口在执行业务前必须过 guard(role, capability)，
 * 越权即抛 AccessDenied。本模块只做授权判定，不持有任何业务逻辑。
 */

export type Role = "core" | "tool" | "gateway" | "ui";

/** 能力清单。每一能力都被映射到一组允许执行它的角色。 */
export type Capability =
  | "engine.decide" // 引擎决策
  | "pipeline.advance" // 阶段链推进
  | "tool.fs_list" // 列出目录
  | "tool.fs_read" // 读取文件
  | "tool.fs_write" // 写入文件
  | "guard.evaluate" // 守卫评估
  | "store.threads" // 会话存储读写
  | "sse.expose" // 暴露 SSE 流
  | "prompts.read"; // 读取烛命提示词

/** 能力 → 允许角色 策略表。 */
const POLICY: Readonly<Record<Capability, readonly Role[]>> = {
  "engine.decide": ["core"],
  "pipeline.advance": ["core"],
  "tool.fs_list": ["core", "tool"],
  "tool.fs_read": ["core", "tool"],
  "tool.fs_write": ["core", "tool"],
  "guard.evaluate": ["core"],
  "store.threads": ["gateway"],
  "sse.expose": ["gateway", "ui"],
  "prompts.read": ["gateway", "ui"],
};

/** 越权异常。 */
export class AccessDenied extends Error {
  constructor(
    readonly role: Role,
    readonly capability: Capability,
  ) {
    super(`access denied: 角色「${role}」无权执行能力「${capability}」`);
    this.name = "AccessDenied";
  }
}

/** 判断某角色是否可执行某能力。 */
export function can(role: Role, capability: Capability): boolean {
  return POLICY[capability].includes(role);
}

/** 守卫生成能力映射的策略表（只读快照，便于审计与测试）。 */
export function capabilityPolicy(): Readonly<Record<string, readonly Role[]>> {
  return POLICY;
}

/** 授权守卫：无权限即抛 AccessDenied。 */
export function guard(role: Role, capability: Capability): void {
  if (!can(role, capability)) {
    throw new AccessDenied(role, capability);
  }
}